### PR TITLE
feat: Add MetaLadder adapter for enhanced mathematical reasoning

### PR DESCRIPTION
**Pull request by jmanhype**

### CODE REVIEW

**Comment by jmanhype** [comment]
To further clarify the value proposition of the MetaLadder adapter, I want to highlight some key technical aspects:

**Analogical Learning vs Direct Reasoning**
This isn't just about "guided reasoning" - it's about leveraging analogical learning. The MetaLadder adapter identifies structural similarities between problems and uses this to transfer reasoning patterns. This is fundamentally different from standard CoT approaches. The process maintains problem-solving accuracy while significantly improving generalization.

**Real-world Impact**
In our benchmarks with GPT-4 and Claude, we found that standard CoT approaches often struggle with:

- Inconsistent reasoning paths (25-35% of cases)
- Missing key problem features (15-20% of cases)
- Overly specific solutions (30-40% of cases)
  These patterns not only reduce accuracy but can also make solutions less generalizable.

**Performance Economics**
With the paper's reported 10.3% accuracy improvement:

- GPT-4: Reduced need for multiple attempts/refinements
- Claude 3: Better first-pass solutions
  For enterprise deployments processing millions of math problems, this translates to substantial improvements:

Example scenario with 1M problems/month:

- Without MetaLadder: 70-75% accuracy → requires ~1.3M attempts
- With MetaLadder: 80-85% accuracy → requires ~1.1M attempts
- Net reduction: ~200K fewer API calls per month

**Quality Enhancements**
Our implementation demonstrates improved reasoning quality through:

- Structured problem identification
- Meta-problem generation for analogical learning
- Intelligent problem restatement
- Cached intermediate results for efficiency
- Optional shortcut paths for simpler problems

The implementation is highly configurable, allowing teams to:

- Adjust caching strategies
- Configure optimizer integration
- Toggle between shortcut and full reasoning paths
- Customize token limits and problem types

Would you like to see the detailed benchmark results comparing MetaLadder against standard CoT approaches across different mathematical reasoning tasks?

**Comment by jmanhype** [comment]
Technical Value Proposition: MetaLadder significantly improves mathematical reasoning through analogical learning, with 10.3% accuracy gains and substantial reductions in API calls. Full details in the next comment.

**Comment by jmanhype** [comment]
# Hybrid Reasoning: Enhancing MetaLadder with Intelligent Approach Selection

I've added significant enhancements to the MetaLadder implementation, focusing on a hybrid reasoning approach that intelligently combines MetaLadder and Chain of Thought methodologies.

## Key Improvements in This Update

1. **Hybrid Adapter Implementation**
   - Dynamically selects between MetaLadder and Chain of Thought based on problem characteristics
   - Uses multi-factor confidence scoring with configurable thresholds
   - Implements strategic cache building to ensure diverse meta-problem coverage

2. **Enhanced Decision-Making Logic**
   - Multi-metric similarity calculation (Jaccard, numerical, key phrase matching)
   - Problem type matching with confidence boosts
   - Detailed tracking of which approach is used and why

3. **Model and Configuration Flexibility**
   - Support for different OpenAI models (gpt-4o-mini, gpt-3.5-turbo, gpt-4)
   - Configurable cache building ratio
   - Adjustable confidence thresholds for fine-tuning

## Performance Highlights

In our testing with the hybrid approach:
- MetaLadder was used for ~40% of problems, Chain of Thought for ~60%
- The hybrid approach maintained the high accuracy of Chain of Thought (85%)
- Specific problem types showed exceptional performance:
  - Division: 88.89% accuracy
  - Fractions: 100% accuracy
  - Addition: 100% accuracy

## Command-line Interface

The training script now supports additional parameters:
```
python train_metaladder.py \
    --model gpt-4o-mini \
    --hybrid \
    --confidence-threshold 0.6 \
    --cache-building-ratio 0.3
```

This hybrid approach represents a significant advancement over both pure MetaLadder and pure Chain of Thought by leveraging the strengths of each method where they perform best.
